include/lbann/models/model.hpp

```diff
@@ -446,7 +449,7 @@ class model {
   std::vector<metric*> m_metrics;
 
   /** @brief Current callbacks to process. */
-  std::vector<callback_base*> m_callbacks;
+  std::vector<std::shared_ptr<callback_base>> m_callbacks;
```

Comment:
The first counterexample I can think of is: a trainer holding multiple models might assign different callbacks to each. In this case, the model can track its own callbacks or the trainer needs a more complicated mapping structure to assign callbacks to models. In the end, I don't know how much it matters, but this seems a reasonable approach for now.

Reply:
Good point that it returns a reference, not a copy. In any case, I don't think we'd lose much performance by copying the `shared_ptr`s into raw pointers. This isn't happening in an inner loop, so the gains in clarity and safety are worth the cost of allocating some memory on the heap.

Reply:
I think the fundamental problem is we have two distinct objects, trainer callbacks and model callbacks, which are shoehorned into the same class. The mapping structure is inherent in the callbacks, so we either need to implement it in the class structure or in the callback implementations.

Thinking about it, many of the "model callbacks" are logically layer callbacks or weights callbacks. We implement the mapping structure internally so that they are applied to the right layer or weights objects.